### ISSUE TITLE
FieldsBuilder not initialize fields, resulting in incorrect gradients

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#4016]
**Describe the bug**
`FieldsBuilder` does not initialize recycled memory, resulting in leaving garbage in `place`d fields, which is not desirable when we need to `place` a `field.grad`.

**To Reproduce**
Try this code with and without `*.grad.fill(0.)`

```python
@ti.kernel
def mul(arr: ti.template(), out: ti.template()):
    for i in arr:
        out[i] = arr[i] * 2.0


@ti.kernel
def calc_loss(arr: ti.template(), loss: ti.template()):
    for i in arr:
        loss[None] += arr[i]


def test_ti_uninitialized_memory():
    ti.init(ti.cpu)
    for i in range(10):
        print(i)
        arr = ti.field(ti.f32)
        fb0 = ti.FieldsBuilder()
        fb0.dense(ti.i, 10).place(arr, arr.grad)
        snode0 = fb0.finalize()
        # arr.grad.fill(0.)
        out = ti.field(ti.f32)
        fb1 = ti.FieldsBuilder()
        fb1.dense(ti.i, 10).place(out, out.grad)
        snode1 = fb1.finalize()
        # out.grad.fill(0.)
        loss = ti.field(ti.f32)
        fb2 = ti.FieldsBuilder()
        fb2.place(loss, loss.grad)
        snode2 = fb2.finalize()
        # loss.grad.fill(0.)
        arr.fill(1.0)
        mul(arr, out)
        calc_loss(out, loss)
        loss.grad[None] = 1.0
        calc_loss.grad(out, loss)
        mul.grad(arr, out)
        assert torch.allclose(torch.ones(10) * 2, out.to_torch()), f"{out.to_torch()}"
        assert torch.allclose(torch.ones(10) * 2, arr.grad.to_torch()), f"{arr.grad.to_torch()}"
        snode0.destroy()
        snode1.destroy()
        snode2.destroy()
```

**Log/Screenshots**
There are simply assertion errors, which means the gradients are not correct.

**Additional comments**
Taichi = 0.8.9

Related issue #4015 in that we need to manually place `.grad`.

I encountered this problem when I develop [stannum](https://github.com/ifsheldon/stannum). It's super subtle and I didn't see any documentation mentioning this issue.

**Comment by k-ye** [comment]
Hi, would you like to submit a PR to fix this?

**Comment by ifsheldon** [comment]
Update:

I tried the below code, but it does not work because the field is not finalized by FieldsBuilder.

```python
# in taichi.lang.snode.SNode.place
    def place(self, *args, offset=None, shared_exponent=False):
        """Places a list of Taichi fields under the `self` container.

        Args:
            *args (List[ti.field]): A list of Taichi fields to place.
            offset (Union[Number, tuple[Number]]): Offset of the field domain.
            shared_exponent (bool): Only useful for quant types.

        Returns:
            The `self` container.
        """
        if offset is None:
            offset = ()
        if isinstance(offset, numbers.Number):
            offset = (offset,)
        if shared_exponent:
            self.ptr.begin_shared_exp_placement()

        for arg in args:
            if isinstance(arg, Field):
                for var in arg.get_field_members():
                    self.ptr.place(var.ptr, offset)
                arg.fill(0)
            elif isinstance(arg, list):
                for x in arg:
                    self.place(x, offset=offset)
            else:
                raise ValueError(f'{arg} cannot be placed')
        if shared_exponent:
            self.ptr.end_shared_exp_placement()
        return self
```
Then the finalization should initialize all fields with `0`, which goes into CPP scope, but I don't know much about the CPP code of Taichi.

**Comment by lin-hitonami** [comment]
Fixed in #4463.
